Surfacing VMIs in the UI (#310)
openshift/openshift-origin-design: web-console/knikubevirt/VMIs/VMIs.md

2. On Bare metal machines installed with Kubernetes or users who install Kubevirt for CNV - On these VMs there are clusters that can be scaled up and down and they create the VMIs (and not the VMs).

## 2 Types of personas

1. Traditional VM operation admin that is used to a certain view and behavior of the console, in a traditional virtualization way/ A user who has a virtualization background and has no need to understand the underlying k8s technology
2. Users that use the console in a more ‘Kubernetes’ manner
We need to try to educate users and narrow the gap for CNV users to better understand K8s/OpenShift.
We want to allow this to be surfaced in the UI and live side by side in a way that will not confuse/ frighten/ intimidate the traditional user.

## Mockups

We had finalized into this option and will show 2 parallel columns in the VMs list page, one for each: VMs beside VMIs.

> ![list of VMIs beside VMs](img/VMsListW_VMIsOp1.png)

Reviewer: Nit: in the image, none running VMs do not have a "node" or "ip" should be empty or some notice.

# Surfacing VMIs in the UI

A virtual machine instance(VMI) is the representation of a running virtual machine.

## Use Cases

1. Debugging -  Previously if a user was trying to resolve an issue caused by a VMI that didn't have an owner it would not be possible to discover it in the UI. Including orphaned VMIs indicates to a user that there could be an issue that they need to address.

2. On Bare metal machines installed with Kubernetes or users who install Kubevirt for CNV - On these VMs there are clusters that can be scaled up and down and they create the VMIs (and not the VMs).

## Mockups

In order to avoid duplication and unclutter the UI, we had finalized into an option that presents VMs and VMIs in a single list view.
We don't surface VMIs in the list if they are owned by a VM.

VMIs and VMs will be distinguished between different color badges. VMI badges color are darker (#002F5D) in order to provide proper contrast for low vision users.

> ![single list view of VMs and VMIs ](img/VMsListW_VMIsOp2.png)

Reviewer: Note: The `HintBlock` component used in the console ( for example in the developer view => +Add page ) does not support closing option.

Reviewer 2: ah, well I know there is active work around updating the hint pattern so we will see it eventually 🤞

## Hints

To better educate users about VMIs, we will provide hints, whenever it is appropriate.
The instance name link will take the user to the instance page (VM/VMI).

![showing hint pattern to help educate users about VMIs](img/Op2_Hint.png)

## VMI's details view

VMI and VM in the list view will lead to the VM and VMI details page.
The details page link from VMI to VM (or note that there is no managing entity=owner) and from VM to VMI (or note that there is no VMI).

VMI's details page is similar to the VM’s details page and acts in a complementary manner.

The VMI's details page will provide a hint with a general explanation about what VMIs are as well as allow drill down to further details.

![VMI's details page](img/Op1_VMI_DetailsViewPlusHint.png)

In the VMI’s details page, the owner VM will be represented with a link, so users can click and get there and vice versa.

## Adding a link to the VMI from the VM details page

In order to let the user get to the VMI even when there is a VM around it, we added a link in the VM details page.

> ![VMI's details page showing link to its VMI](img/VM_DetailsPageW_LinkToVMI.png)

Reviewer: a - In this image we have a link to the pod, IIRC we decided to remove this link on Jan 9 mitting, @matthewcarleton please correct me if I'm wrong.

b - In this image the link to VMI is in the same place as owner ... this is very confusing as the VM is the owner of VMI and not the other way around. IIRC we decided to replace the POD link with a VMI link (~top right area) and leave the owner link for the VM owner link (bottom left) as it is now.

Reviewer 2: @yaacov ya I think you're right. I believe Fabian said that. 
I'm a bit confused, this makes sense to me. If I look at the VMI I see the owner VM if I look at the VM I see the VMI. Maybe I'm not understanding correctly.

Reviewer: > I'm a bit confused, this makes sense to me. If I look at the VMI I see the owner VM if I look at the VM I see the VMI. Maybe I'm not understanding correctly.

VM is the manager of VMI ( e.g. owner )
This is a [k8s thing](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents) [1] and apply to many resources, in the UI we usually display the owner of a resource in the bottom left. 

VMI is a resource that VM mange ( e.g. dependent ):
a - If their are multiple resources - we currently display them as a list at the buttom of the page, for example containers and volumes in the Pod details page.
b - If their is only one, or it's important to show in the top section, we add it to the details on top, like we currently do for Pod in VM details view.

[1] https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents

Reviewer 2: and you are proposing we keep the owner for the VM? Is this where we'd possibly see VMIRS in the future? My concern is showing something that is empty all the time, which could lead to confusion.

## Actions available for the VMI

> ![VMI's available actions](img/Op1Actions.png)

Reviewer: @matthewcarleton  hi, do we have an example of other resource that use the info banner above the details section ?
( It's will be very helpful to have an implementation reference )

> ![VMI's available actions](img/Op1Actions.png)

Reviewer: In our design we only show VMI if they have no VM, so users will not have access to a VMI details page if it has a VM that mask it.

@matthewcarleton @yfrimanm  do we want to let user access the "VM-like" details page of a VMI that has a VM ( a. it will be a duplication of the VM details view, b. we do not link to it from the list view ) ? 
WDYT ?

Author: @yaacov AFAIR we said we want users to interact with VMs, so IMO, we should allow that.

Reviewer 2: @yaacov what happens if I go to the details page of a VM and click on the instance there? Won't I go to the VMI details page?

Reviewer: > @yaacov what happens if I go to the details page of a VM and click on the instance there? Won't I go to the VMI details page?

in https://github.com/openshift/console/pull/3949 **currenlty WIP** you will get to an actuall "Virtual Machine Instanse" page, that is a generic page for VMI , not the "Virtual Machine" page, with the dashboards, metrics and events.

Reviewer 2: ok, I'd vote for the same experience regardless of how you get to the VMI (list view as an orphan or via the VM details page.) Does that make sense?

Reviewer: > ok, I'd vote for the same experience regardless of how you get to the VMI (list view as an orphan or via the VM details page.) 

I have a technical problem:
the VMI and VM use the same details page, `../ns/<namesapce>/virtualmachines/<name>/` their is no difference between a VM and a VMI, like the list the details page check if we have a vm, vmi or vm+vmi, it renders vm page if we have vm, and only if we do not have a vm it renders the vmi view/actions

> Does that make sense?

sure, but I think the solution will be to remove the link ... or make vm and vmi a different page, e.g. different URL and different breadcrumbs

Reviewer 2: ah I see what you're saying @yaacov So we don't want to surface VMI in the VM b/c it adds no additional value. We only show VMI details pages when the VM isn't present. That makes sense to me. @yfrimanm WDYT?

Reviewer: yes, the vm details page is actually "vm+vmi" details page, when collecting the data to show in the page we take all the data we have from the vm + all the data we have from the vmi, and show a combined view of all the data we have. Linking from the vm details page to the vmi details page is actually linking to the same page if It make sense :-)

Author: I see the tricky issue here. So I think we might need to cancel that link. WDYT?

Reviewer: > I see the tricky issue here. So I think we might need to cancel that link. WDYT?

looks like it should be a link to pod ( like it was in the beginning ) :-)

Reviewer 2: ya I think that aligns with how we present it in the list view.
VMI only present if it exists on its own. Otherwise you interact with the VM. I like that more actually b/c then we would surface just the Pod from the VM. 
@yfrimanm can you update the designs to reflect this?

4. Upstream Kubernetes is popular because in VMI Replica set and machine controller implementation that is used to “spork(?)” Kubernetes clusters on top of CNV
Bare metal machines installed with Kubernetes
Users install Kubevirt for CNV
On these VMs there are clusters that can be scaled up and down and they create the VMIs (and not the VMs)

## 2 Types of personas

1. Traditional VM operation admin that is used to a certain view and behavior of the console, in a traditional virtualization way/ A user who has a virtualization background and has no need to understand the underlying k8s technology
2. Users that use the console in a more ‘Kubernetes’ manner
We need to try to educate users and narrow the gap for CNV users to better understand K8s/OpenShift.
We want to allow this to be surfaced in the UI and live side by side in a way that will not confuse/ frighten/ intimidate the traditional user.

## Mockups

We had finalized into this option and will show 2 parallel columns in the VMs list page, one for each: VMs beside VMIs.

> ![list of VMIs beside VMs](img/VMsListW_VMIsOp1.png)

Reviewer: a - On none running machines we do not have "created", "node" or "ip" columns,
What we should show on the missing columns ?

b - +1 for renaming "Name" and "Instance"

> ![list of VMIs beside VMs](img/VMsListW_VMIsOp1.png)

Author: Thanks Andy! 🙂 +1X2

Trying to educate users about VMIs, we will provide hints, whenever it is appropriate. 
In most usual cases, when no VM or VMI is missing, the instance name link will take the user directly to the instance page (VM/VMI).

> When a VMI is missing an owner VM, the user will be provided with an explanation in a popover for the reason it is missing. The name of the VM + badge will be shown in a disabled state for the sake of accessibility so a screen reader would be able to announce it.

Author: @andybraren thanks! +1 on "No instance".

> ![VMI is in an off status](img/Op1_Hint1.png)

Reviewer: > but as far as I know this would be the first time that clicking a resource's blue text would sometimes reveal a popover instead of immediately taking the user to that resource's Details page.

+1 , also on rows where we have VMs the blue text will immediately take the user to that resource's Details page, making two different behaviour patterns on the same list.

Maybe we can ?
1 - Do not link to VMIs if we have a VM, because if we have a VM user do need to interact with the VMI directly.
     Users can get to the VMI using the VM's detail page. 
2 - Show the VMIs that are owned by VMIRS ( e.g. "Orphan VMI" ) in a separate page.

> ![VMI is in an off status](img/Op1_Hint1.png)

Author: +1 
Updated the popover to connect with the greyed "No VM" and not to the blue resource.

### UI definitions for missing VMs and VMIs in the table in a disabled state

Color of badge: #737679, background color of badge: #D2D2D2, color of name text: #737679

If a VMI is in an off status, the user will also be notified about that via a popover.

> ![Missing VMI](img/Op1_PopoverMessageMissingVMI.png)

Reviewer: +1 for "No instace"

> ![Missing VMI](img/Op1_PopoverMessageMissingVMI.png)

Author: +1 @lizsurette, thanks.

## VMI's details view

VMI's details page should be similar to the VM’s details page and act in a complementary manner.
The VMI's details page will provide a hint with a general explanation about what are VMIs and also allow drill down to further details.

> ![VMI's details page](img/Op1_VMI_DetailsViewPlusHint.png)

Author: +1 
Thanks @andybraren

> ![VMI's details page](img/Op1_VMI_DetailsViewPlusHint.png)

Author: > Does it mean that any edits they make to the VMI on this page will be overwritten by the VM:

Kobi explains that VMIs Like Pods will change if their owner changes, so if this VMI has an owner (VM or VMIRS), they will change on the next boot to reconcile with their owner.

> Do the actions in the Actions menu on this page apply to the VM instead of the VMI?
No, these actions apply to the VMI's.

> ![VMI's details page](img/Op1_VMI_DetailsViewPlusHint.png)

Author: +1 @lizsurette thanks!

> In the VMI’s details page, the owner VM will be represented with a Iink, so users can click and get to the owner VM details page and vice versa. The term ‘owner VM’ will be furthered explained in a popover.

Reviewer: +1 for "Owner"  to align with all other resources.

We might want to surface which VMI doesn’t have an owner VM (orphan VMI) via a notification or maybe even in the dashboard overview page.

## Actions available for the VMI

> ![VMI's available actions](img/Op1Actions.png)

Author: @andybraren AFAIK, the actions apply to the VMI but we better advise @rmohr

> If the user chooses to delete a VMI, they will be notified that this action will also delete the owner VM of that VMI.

Author: OK, I was mixing it up. Sorry. 
If the user deletes a managed VMI then the owner will restart the VMI again.

> ![Notifying the user about deleting the VMI will delete its VM](img/Op1WarningPopupDelete.png)

Author: Thanks @andybraren!

# Surfacing VMIs in the UI

> VMI is the current running VM.

Reviewer 2: ```suggestion
A virtual machine instance(VMI) is the representation of a running virtual machine.
```

> The initial reason for surfacing VMIs in the UI was that there is no point in hiding the VMIs from the UI.

Reviewer 2: I think we can remove this line since you get into the use cases below.

> ## UCs

Reviewer 2: nitpic:
```suggestion
## Use Cases
```

> 1. Debugging - If a user thinks that something isn’t right with the VMs and he wants to check what’s going on. When a VMI isn’t managed by a VM and it’s invisible in the UI but that VMI is actually still there because it’s impossible to delete VMIs (that’s a basic Kubernetes mechanism). This can happen by an orphan delete, or that a replica set isn’t used.

Reviewer 2: ```suggestion
1. Debugging - Previously if a user was trying to resolve an issue caused by a VMI that didn't have an owner it would not be possible to discover it in the UI. Including orphaned VMIs indicates to a user that there could be an issue that they need to address.
```

2. On Bare metal machines installed with Kubernetes or users who install Kubevirt for CNV - On these VMs there are clusters that can be scaled up and down and they create the VMIs (and not the VMs).

> ## 2 Types of personas

Reviewer 2: We haven't included personas in the repo in the past have we? Maybe we shouldn't here. WDYT?

1. Traditional VM operation admin that is used to a certain view and behavior of the console, in a traditional virtualization way/ A user who has a virtualization background and has no need to understand the underlying k8s technology
2. Users that use the console in a more ‘Kubernetes’ manner
We need to try to educate users and narrow the gap for CNV users to better understand K8s/OpenShift.
We want to allow this to be surfaced in the UI and live side by side in a way that will not confuse/ frighten/ intimidate the traditional user.

## Mockups

In order to avoid duplication and unclutter the UI, We had finalized into an option that presents  VMs and VMIs in a single list view.
We don't surface VMIs in the list if they are owned by a VM.

> VMIs and VMS will be distinguished between different color badges. VMI badges color will be bolder than VMs badges color, for accessibility reasons (#002F5D).

Reviewer 2: ```suggestion
VMIs and VMs will be distinguished between different color badges. VMI badges color are darker (#002F5D) in order to provide proper contrast for low vision users.
```

![single list view of VMs and VMIs ](img/VMsListW_VMIsOp2.png)

## Hints

> Trying to educate users about VMIs, we will provide hints, whenever it is appropriate.

Reviewer 2: ```suggestion
To better educate users about VMIs, we will provide hints, whenever it is appropriate.
```

> In most usual cases, the instance name link will take the user directly to the instance page (VM/VMI).

Reviewer 2: Wouldn't this always happen?

> ![showing hint pattern to help educate users about VMIs](img/Op2_Hint.png)

Reviewer 2: I wonder if we should state in the hint why a VMI would show up in the list view.

Author: Yes. Just added "The VMIs listed below are not owned by a VM".

Reviewer 2: nitpic: you have "not owned by a virtual machines"

Reviewer 2: @jelkosz would you say this text is accurate from a dev perspective?

Author: @matthewcarleton changed to: 'The Virtual Machine Instances listed below are not owned by Virtual Machines'. 
Maybe: 'The Virtual Machine Instances listed below don't have Virtual Machines owners' sounds better?

Reviewer 2: +1 that works too!

## VMI's details view

VMI and VM in the list view will lead to the VM and VMI details page.
The details page link from VMI to VM (or note that there is no managing entity=owner) and from VM to VMI (or note that there is no VMI).

VMI's details page is similar to the VM’s details page and acts in a complementary manner.

> The VMI's details page will provide a hint with a general explanation about what are VMIs and also allow drill down to further details.

Reviewer 2: ```suggestion
The VMI's details page will provide a hint with a general explanation about what VMIs are as well as allow drill down to further details.
```

![VMI's details page](img/Op1_VMI_DetailsViewPlusHint.png)

In the VMI’s details page, the owner VM will be represented with a link, so users can click and get there and vice versa.
The term ‘Owner’ will be further explained in a popover.

> We chose to use the term ‘Owner’ to align with the current phrasing along the console.

Reviewer 2: I think we can remove this bit from the repo, just good for us to know if anyone asks further about it :)
